### ISSUE TITLE
basecalling very slowly when training my own model

### ISSUE THREAD

**Issue by defendant602** [nanoporetech/bonito#91]
Hi:
          I want to train my own model using bonito v0.3.2, so first I need to basecall my training set using default model dna_r9.4.1. But I found the basecalling process output very slowly. When I checked the log info, I found i got an memory issue `OSError: [Errno 12] Cannot allocate memory`
     my command is
 `bonito basecaller dna_r9.4.1 --save-ctc --reference ref.fasta ./training_set/ > ./training/ctc-data/basecalls.sam 2> basecall.log`

tailing info in basecall.log

> Traceback (most recent call last):
>   File "/root/Miniconda/envs/bonito/lib/python3.6/threading.py", line 916, in _bootstrap_inner
>     self.run()
>   File "/root/Miniconda/envs/bonito/lib/python3.6/threading.py", line 864, in run
>     self._target(*self._args, **self._kwargs)
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/pool.py", line 405, in _handle_workers
>     pool._maintain_pool()
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/pool.py", line 246, in _maintain_pool
>     self._repopulate_pool()
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/pool.py", line 239, in _repopulate_pool
>     w.start()
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/process.py", line 105, in start
>     self._popen = self._Popen(self)
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/context.py", line 277, in _Popen
>     return Popen(process_obj)
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/popen_fork.py", line 19, in __init__
>     self._launch(process_obj)
>   File "/root/Miniconda/envs/bonito/lib/python3.6/multiprocessing/popen_fork.py", line 66, in _launch
>     self.pid = os.fork()
> OSError: [Errno 12] Cannot allocate memory
> calling: 18775 reads [4:59:05, 196.01s/ reads]

this is the output of `nvidia-smi`

> +-----------------------------------------------------------------------------+
> | NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
> |-------------------------------+----------------------+----------------------+
> | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
> |===============================+======================+======================|
> |   0  Tesla V100-SXM2...  Off  | 00000000:21:01.0 Off |                    0 |
> | N/A   29C    P0    37W / 300W |   5770MiB / 16160MiB |      0%      Default |
> +-------------------------------+----------------------+----------------------+
>                                                                                
> +-----------------------------------------------------------------------------+
> | Processes:                                                       GPU Memory |
> |  GPU       PID   Type   Process name                             Usage      |
> |=============================================================================|
> |    0     29180      C   /root/Miniconda/envs/bonito/bin/python3.6   5641MiB |
> +-----------------------------------------------------------------------------+

I ran bonito on a server with 8C 64G, after the OSError occured, the basecalling process did not stop, but output very slowly.

Could anyone help me on this issue? Thanks! and sorry about my text formatting.

**Comment by iiSeymour** [comment]
Hey @defendant602 

When using `--save-ctc` all the training data is accumulated in memory and only written to disk at the end of the process. 

Unfortunately, it appears you have run out of system memory which leaves two options: 
 - Use a smaller subset of reads.
 - Use a node with more system RAM if possible.

HTH

Chris.

**Comment by defendant602** [comment]
Thanks for your quick help and reply!

I downsampled my training set and used another node with more RAM, I worked just fine by now.

A small question by the way, I noticed that in the output sam file `training/ctc-data/basecalls.sam`, the reads IDs are formatted like "0000c5b0-45b6-4bcb-a053-5a58e0fb6f14:1:36", "0000c5b0-45b6-4bcb-a053-5a58e0fb6f14:2:36", "0000c5b0-45b6-4bcb-a053-5a58e0fb6f14:3:36" ,etc. Is it normal? I assume the right formatted reads ID should be "0000c5b0-45b6-4bcb-a053-5a58e0fb6f14"?

If it is normal, what do the numbers mean after the regular reads ID?
